esp-hal/src/dma/mod.rs

```diff
@@ -733,58 +727,72 @@ where
 
     fn available(&mut self) -> usize {
         if self.last_seen_handled_descriptor_ptr.is_null() {
-            self.last_seen_handled_descriptor_ptr = self.descriptors.as_mut_ptr();
-            return 0;
-        }
-
-        if self.available != 0 {
-            return self.available;
+            // initially start at last descriptor (so that next will be the first
+            // descriptor)
+            self.last_seen_handled_descriptor_ptr =
+                addr_of_mut!(self.descriptors[self.descriptors.len() - 1]);
         }
 
-        let descr_address = self.last_seen_handled_descriptor_ptr;
-        let mut dw0 = unsafe { descr_address.read_volatile() };
-
-        if dw0.owner() == Owner::Cpu && !dw0.is_empty() {
-            let descriptor_buffer = dw0.buffer;
-            let next_descriptor = dw0.next;
-
-            self.read_buffer_start = descriptor_buffer;
-            self.available = dw0.len();
-
-            dw0.set_owner(Owner::Dma);
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        let mut current_in_descr_ptr =
+            unsafe { self.last_seen_handled_descriptor_ptr.read_volatile() }.next;
+        let mut current_in_descr = unsafe { current_in_descr_ptr.read_volatile() };
 
-            unsafe {
-                descr_address.write_volatile(dw0);
-            }
+        while !current_in_descr.is_empty() && current_in_descr.owner() == Owner::Cpu {
```

Comment:
This check isn't needed. `pop` doesn't check it either.
```suggestion
        while current_in_descr.owner() == Owner::Cpu {
```